### ISSUE TITLE
[01] Eval/IO overhead in the greedy demo

### ISSUE THREAD

**Issue by SomeoneSerge** [sinanmohd/evanix#10]
Running `evanix` on https://github.com/ggerganov/llama.cpp takes, in the `--dry-run` mode, about 20s, whereas running `nix-eval-jobs` takes about 6s. Per @sinanmohd's hypothesis, this is due to querying the substituters about the cached store paths too much (does nix not maintain a negative response cache for these queries?). The performance needs to be brought down to the same order of magnitude as `nix-eval-jobs` so that we can stop talking about the greedy PoC

**Comment by sinanmohd** [comment]
the 6s nix-eval-jobs eval is probably because of only evaluating `#packages.x86_64-linux`
`nix-eval-jobs --flake .#packages.x86_64-linux`

you can get a similar performance from evanix with
`evanix --dry-run --flake .#packages.x86_64-linux`

where as `--system x86_64-linux` still evaluates everything in #packages and drops if derivations with mismatched system

**Comment by sinanmohd** [comment]
i tried running evanix & nix-eval-jobs on https://github.com/ggerganov/llama.cpp

`time evanix --dry-run --system x86_64-linux --flake --max-build 555 --pipelined=false --solver-report .#packages

real    0m15.625s
user    0m0.001s
sys     0m0.004s
`

`time nix-eval-jobs --force-recurse --flake .#packages

real    0m15.877s
user    0m7.297s
sys     0m3.474s
`

the time taken should not vary that much

**Comment by SomeoneSerge** [comment]
```
❯ time nix run github:sinanmohd/evanix -- --system x86_64-linux --flake .#legacyPackages.x86_64-linux --dry-run --max-build 50000 > evanix.stdout 2>evanix.stderr
# ...
# waiting but obviously longer than nix-eval-jobs...
# ...
```

```
❯ nix develop github:sinanmohd/evanix --command -- time nix-eval-jobs --flake .#legacyPackages.x86_64-linux >nix-eval-jobs.stdout 2>nix-eval-jobs.stderr
❯ tail nix-eval-jobs.stderr  -n2
200.85 user 88.72 system 9:02.31 elapsed 53% CPU (0avgtext+0avgdata 4434872maxresident)k
447568inputs+253408outputs (1major+12985415minor)pagefaults 0swaps
```

**Comment by sinanmohd** [comment]
master without locality check (all drvs need to built)

```
$ time evanix --flake .#legacyPackages.x86_64-linux --dry-run --max-build 500
...
real    11m38.698s
user    10m11.110s
sys     0m1.050s
```

**Comment by sinanmohd** [comment]
master without locality check (all drvs need to built)

```
$ time evanix --flake .#legacyPackages.x86_64-linux --dry-run --max-build 2147483647

real    5m35.859s
user    0m2.344s
sys     0m0.923s
```

**Comment by SomeoneSerge** [comment]
> without locality check (all drvs need to built)

How do you turn this mode on/off? WHen you turn this on, does any selection algorithm run at all (e.g. the way I imagine this is you'd compute the full closure, and generate an all-1s "value" vector)?

**Comment by sinanmohd** [comment]
the solver is still in use here, i used to have a patch on git stash
now you can use `--cache-status=false` on master

**Comment by SomeoneSerge** [comment]
> the solver is still in use here, i used to have a patch on git stash

👍🏻 but what are the inputs to the solver?

**Comment by sinanmohd** [comment]
yes, if you mean cost vector by value. cache locality check sets cost to 0 for substituter cached derivatives.
but we entirely remove locally cached dependencies from the DAG.

**Comment by SomeoneSerge** [comment]
So when you say "without locality check (all drvs need to built)" you mean `cost=(1, 1, ..., 1)` without zeros?

**Comment by Mic92** [comment]
https://nix.dev/manual/nix/2.23/command-ref/conf-file.html#conf-narinfo-cache-negative-ttl I would expect this setting would be used for caching cache status. You can check 'nix-build -vvvvv --dry-run' to debug in nix if this is the case.

**Comment by sinanmohd** [comment]
> So when you say "without locality check (all drvs need to built)" you mean `cost=(1, 1, ..., 1)` without zeros?

yes

**Comment by sinanmohd** [comment]
yes, if you mean cost vector by "value vector". It's as if we have all-1s cost vector. cache locality check sets cost to 0 for subtitutotr cached derivatives.
but we entirely remove locally cached deprivations from the DAG.

19-Jul-2024 14:01:52 Someone ***@***.***>:
